### DISCUSSION TITLE
How can I increase the speed of scaling

### DISCUSSION THREAD

**Discussion by IphoneCoder** [libvips/libvips#4157] [Q&A]
## I need to scale the rgba data in memory. How can I speed it up

**My Code**
```
VipsImage *source_img = NULL;
vips_image_new_from_memory(src, src_width * src_height * 4,
                                     src_width,
                                     src_height,
                                     4,
                                     VIPS_FORMAT_UCHAR);
source_img->Type = VIPS_INTERPRETATION_sRGB;
VipsImage *resized_image = NULL;
vips_thumbnail_image(source_img, &resized_image,out_width, 
                        "height", out_height, NULL);
size_t size = 0;
void *dst = malloc(out_width*out_height*4);
int ret = vips_rawsave_buffer(resized_image,&dst, &size,NULL);
```
This code works fine, but by documentation, the vips_thumbnail_image method may be slow, but couldn't find a api to speed it up(vips_thumbnail can not load from memory)。
vips_thumbnail_image took very little time to test, but vips_rawsave_buffer took much longer
eg 400x400 -> 100x100
vips_thumbnail_image take less 1ms
vips_rawsave_buffer take 42ms

**Environment**
(please complete the following information)
 - OS: MacOS 14.5/M3 PRO
 - Vips: based on master, -Dbuildtype=release

**Comment by jcupitt** [comment]
Try (untested):

```C
VipsImage *image = vips_image_new_from_memory(src, src_width * src_height * 4,
                                     src_width,
                                     src_height,
                                     4,
                                     VIPS_FORMAT_UCHAR);
image->Type = VIPS_INTERPRETATION_sRGB;

VipsImage *x;
double scale = (double) out_width / image->Xsize;
if (vips_resize(image, &x, scale, "kernel", VIPS_KERNEL_LINEAR, NULL)) {
  g_object_unref(image);
  return -1;
}
g_object_unref(image);
image = x;

size_t size = 0;
void *dst = vips_write_to_memory(image, &size);
g_object_unref(image);
```

I don't know if this will be especially quick. For simple, single operations on small images like this, libraries like skia are probably a better fit.

libvips is a pipelined image processing library, so it'll start to pull ahead if you have a sequence of operations to perform. The longer the pipeline and the larger the images, the better for libvips.
